src/SIS_dyn_cgrid.F90

```diff
@@ -1738,17 +1781,192 @@ subroutine basal_stress_coeff_C(G, mi, ci, sea_lev, CS)
+  pi = 4.0*atan(1.0)
+
+! a note regarding hi_min and hin_max(0):
+! both represent a minimum ice thickness.  hin_max(0) is
+! intended to be used for particular numerical implementations
+! of category conversions in the ice thickness distribution.
+! hi_min is a more general purpose parameter, but is specifically
+! for maintaining stability in the thermodynamics.
+! hin_max(0) = 0.1 m for the delta function itd
+! hin_max(0) = 0.0 m for linear remapping
+!
+! Also note that the upper limit on the thickest category
+! is only used for the linear remapping scheme
+! and it is not a true upper limit on the thickness
+  hin_max(0)=0.0
+  do k=1,nCat
+    hin_max(k) = IG%mH_cat_bound(k)/rho_ice
+  end do
+
+  Tbt=0.0
+
+  do j=jsc-1,jec+1
+    do i=isc-1,iec+1
+
+      acat(1:ncat) = IST%part_size(i,j,1:ncat)
+      atot = sum(acat)
+      vcat(1:ncat) = IST%mH_ice(i,j,1:ncat) / rho_ice
+      m_i = sum(vcat)
+
+      if (atot > 0.05 .and. m_i > CS%basal_stress_min_thick .and. &
+          sea_lev(i,j) < CS%basal_stress_max_depth) then
+
+        mu_b = G%bathyT(i,j) + sea_lev(i,j)         ! (hwater) mean of PDF (normal dist) bathymetry
+        wid_i = CS%basal_stress_max_depth/ncat_i    ! width of ice categories
+        wid_b = 6.0*CS%sigma_b(i,j)/ncat_b          ! width of bathymetry categories (6 sigma_b = 2x3 sigma_b)
+
+        x_k(:) = (/( wid_i*( real(ii) - 0.5 ), ii=1, ncat_i )/)
+        y_n(:) = (/( ( mu_b - 3.0*CS%sigma_b(i,j) ) + (real(ii) - 0.5) * (6.0*CS%sigma_b(i,j)/ncat_b), &
+                         ii=1, ncat_b )/)
+
+        v_i=0.0
+        do n =1, ncat
+          v_i = v_i + vcat(n)**2 / (max(acat(n), CS%puny))
+        enddo
+        v_i = v_i - m_i**2
+
+        ! parameters for the log-normal
+        mu_i    = log(m_i/(CS%onemeter * sqrt(1.0 + v_i/m_i**2)))
+        sigma_i = max(sqrt(log(1.0 + v_i/m_i**2)), CS%puny)
+
+        ! max thickness associated with percentile of log-normal PDF
+        ! x_kmax=x997 was obtained from an optimization procedure (Dupont et al.)
+
+        x_kmax = CS%onemeter * exp(mu_i + sqrt(2.0*sigma_i)*CS%basal_stress_cutoff)
+
+        ! Set x_kmax to hlev of the last category where there is ice
+        ! when there is no ice in the last category
+        cut = x_k(ncat_i)
+        do n = ncat,-1,1
+          if (acat(n) < CS%puny) then
+            cut = hin_max(n-1)
+          else
+            exit
+          endif
+        enddo
+        x_kmax = min(cut, x_kmax)
+
+        g_k(:) = exp(-(log(x_k(:)/CS%onemeter) - mu_i) ** 2 / (2.0 * sigma_i ** 2)) / (x_k(:) * sigma_i * sqrt(2.0 * pi))
+
+        b_n(:)  = exp(-(y_n(:) - mu_b) ** 2 / (2.0 * CS%sigma_b(i,j) ** 2)) / (CS%sigma_b(i,j) * sqrt(2.0 * pi))
+
+        P_x(:) = g_k(:) * wid_i
+        P_y(:) = b_n(:) * wid_b
+
+        do n =1, ncat_i
+          if (x_k(n) > x_kmax) P_x(n)=0.0
+        enddo
+
+        ! calculate Tb factor at t-location
+        do n=1, ncat_i
```

Comment:
This line is hard-coded to run 10,000 times for every model grid cell.  This seems inflexible and possibly excessive?  I think that the values of ncat_i and ncat_b should be something that can be set at run-time.